Extensions/MUC-Sub/XMPPMUCSub.m

```diff
@@ -0,0 +1,699 @@
+    if (nil == items) {
+        return;
+    }
+    
+    // All preconditions show that it's a MUC-Sub message. Extract the original message
+    // and forward it to the registered delegates. The message may contain several <item>
+    // elements and thus several original <message> elements.
+    dispatch_block_t block = ^{ @autoreleasepool {
+        for (NSXMLNode *item in items.children) {
+            NSXMLNode* presenceNode = [item childAtIndex:0];
+            if (nil == presenceNode) {
+                continue;
+            }
+            
+            if (NSXMLElementKind != presenceNode.kind) {
+                continue;
+            }
+            
+            XMPPPresence *p = [XMPPPresence presenceFromElement:(NSXMLElement *)presenceNode];
+            [multicastDelegate xmppStream:sender didReceivePresence:p];
+        }
+    }};
+    
+    if (dispatch_get_specific(moduleQueueTag)) {
+        block();
+    }
+    else {
+        dispatch_async(moduleQueue, block);
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+#pragma mark -
+#pragma mark Helpers
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+- (NSError *)errorFromIQ:(XMPPIQ *)iq
+{
+    // An error may look like this.
+    // 
+    // <iq xmlns="jabber:client" 
+    //      lang="de" 
+    //        to="hag66@shakespeare.example" 
+    //      from="coven@muc.shakespeare.example" 
+    //      type="error" 
+    //        id="23EAAB2B-F6BD-4CA0-9539-DAE495ED5885">
+    //   <subscriptions xmlns="urn:xmpp:mucsub:0"/>
+    //   <error code="403" type="auth">
+    //     <forbidden xmlns="urn:ietf:params:xml:ns:xmpp-stanzas"/>
+    //     <text xmlns="urn:ietf:params:xml:ns:xmpp-stanzas" lang="de">
+    //       Moderatorrechte benötigt
+    //     </text>
+    //   </error>
+    // </iq>
+    
+    if ([iq isErrorIQ]) {
+        NSXMLElement* error = [iq childErrorElement];
+        
+        NSString *reason = nil;      // Must be filled.
+        NSString *description = nil; // May be filled.
+        
+        for (NSXMLNode *child in error.children) {
+            // If there is a <text> node we can get a useful description. Otherwise
+            // we have to live with the generic <forbidden> (from the example above).
+            // Errors types may vary, of course.
+            if ([child.name isEqualToString:@"text"]) {
+                description = child.stringValue;
+            }
+            else {
+                reason = child.name;
+            }
+        }
+        
+        return [[NSError alloc] initWithDomain:XMPPStreamErrorDomain 
+                                          code:XMPPStreamInvalidState 
+                                      userInfo:@{reason: description}];
+    }
+    
+    return nil;
+}
+
+
+- (NSArray<XMPPJID *>*)jidFromSubscriptionIQ:(XMPPIQ *)iq
```

Comment:
Whose JID?

Should headers/signatures for these helpers be above in `@interface XMPPMUCSub (PrivateAPI)`?

Reply:
Depends on the context. This is the structure that is parsed. 
```xml
<iq from='muc.shakespeare.example'
    to='hag66@shakespeare.example'
    type='result'
    id='E6E10350-76CF-40C6-B91B-1EA08C332FC7'>
  <subscriptions xmlns='urn:xmpp:mucsub:0'>
    <subscription jid='coven@muc.shakespeare.example' />
    <subscription jid='chat@muc.shakespeare.example' />
  </subscriptions>
</iq>
```
The < subscriptions > tags are read and either mean subscribed users in a room or room a user is subscribed to.

I removed the `(PrivateAPI)` extension because it's not necessary.

Reply:
Added structure as comment at the top of the function.